### DISCUSSION TITLE
Multi-slice custom input?

### DISCUSSION THREAD

**Discussion by ckoehler** [winnow-rs/winnow#528] [Q&A] [answered]
Hey there! 

Our main input type (incidentally also called `Stream`) is backed by a Vec of `Packet` that all have data in backing slices. I.e. there isn't one contiguous memory region of data in our Stream (only in each `Packet`). I think that means we can't implement `AsBytes`, and our `Slice` type for `winnow::stream::Stream` can't just be `&[u8]`, etc. That makes it impossible to use Winnow, at least if we insist on using our `Stream`.

We could implement (Winnow's) Stream on our `Packet` type, and I think that may work, but I was wondering if you have any other ideas we may pursue.

Thanks!

Christoph

**Comment by epage** [comment]
nom v7 had more individual traits, letting parsers declare what their requirements are.

This made it more difficult to make custom streams, more difficult to make generic parsers, and makes the parsers brittle where subtle implementation changes can change the requirements on the traits.  We resolved this by combining almost all of the traits into `Stream`.

Even if we split out contiguous and token-focused streams, I wonder if there would be enough functionality left to be of us.  If there is enough of a case for token-only parsing to be sufficient, we can revisit the trait design.

Thoughts?

**Reply by ckoehler** [reply]
To clarify, our `Token`s are `u8`, too, they just happen to be stored in non-contiguous memory. We deal mostly with binary data.

The particular problem I'm having is that Winnow parses numbers differently than nom, requiring `AsBytes` on the `Slice`, whereas nom iterates and depends on `AsChar` for the iterator `Item` (`u8` in our case) to parse numbers. That works for our Stream.

Don't take our use case as normative, it may not be. I just wanted to throw it out there to see if there are ideas or similar stories. And maybe the right answer is to parse our `Packet`s that are contiguous instead of the whole Stream. 🤷‍♂️ 

Thanks for replying so quickly!

**Reply by ckoehler** [reply]
Would it be possible to have AsBytes return a Cow? Something like:

```rust
impl<'a> AsBytes for MyStream<'a> {
    fn as_bytes(&self) -> Cow<'_, [u8]> {
        self.to_slice() // a method we already have that returns a Cow
    }
}
```

That would let us implement this, and we'd only have to copy bytes if the requested data spans multiple packets internally. With the current signature, even if we convert to a Cow, the data would be owned by `as_bytes` (since it first copies into an owned vec if needed), so it'd get dropped at the end of the function and Rust complains (you understand this, just giving more context for others reading it).

For existing users who already have just a slice, that'd require wrapping it in a `Cow::Borrowed`, which should be pretty cheap (though a bit less readable).

Again, maybe our use case just means we're odd and can't use Winnow, and "no" would be a perfectly acceptable answer! I know it's a lot to ask. Also open to other ideas!

**Reply by epage** [reply]
This runs into more fundamental problems than `AsBytes::as_bytes`.  For example, we first call `digit0` to validate the format.  That calls `take_while` which requires returning a contiguous slice.

More fundamental than that is that `Stream` requires being able to return contiguous slices.

As I mentioned, this would require splitting up `Stream` into at least a `TokenStream` and a `SliceStream`.  We'd then need a good feel for what parsers should be adapted to only work with a `TokenStream`.    Depending on the performance characteristics (if performance matters for some of these), we then might have to keep some of those parsers as requiring `SliceStream`s.  How much is left and how useful that is would help us to know whether going in this direction is worth it.

**Reply by ckoehler** [reply]
Thanks Ed! I'm sorry I haven't been very clear, it's hard when I can't share the code... 😞 

I don't think it's necessary to split `Stream` up. We already implemented enough of it in such a way that `digit0` and `take` and `take_while` etc work just fine, right now. We do that by setting the associated type `Slice` to also be a `MyStream` (and I can elaborate on the implementation if you're interested).

But for `AsBytes`, we have no choice but to return a `&[u8]` because of the signature. And even though we _can_ turn a `MyStream` into a `&[u8]` via `Cow`, that doesn't help with `AsBytes` because the data is not stored contiguously and has to be copied first. That makes `as_bytes` the owner of the now-contiguous data, and Rust rules won't let us return a slice to that because the owned data will get dropped after return.

Some brief internal tests to change `AsBytes` to return a `Cow` instead seem to work for our use case (patch: [winnow.patch](https://github.com/user-attachments/files/15536626/winnow.patch)).

I hope this explains our use case better. This still doesn't solve the breaking change to a trait and its consequences, which may still be a deal breaker. But hopefully that makes our situation clearer; maybe I'm missing a better answer altogether.

Thanks again for engaging!

**Reply by epage** [reply, marked as answer]
Thanks for the clarification!

I am curious what this `Stream` looks like but eh, if you have it working.

I'd prefer `AsBytes` to not return a `Cow`.

The current implementation of the binary number parsers (the part I'm assuming you are running into issues with) were simplified to use `take` which returns `Stream::Slice`.  Processing the `Stream::Slice` is an issue and we used `AsBytes` to simplify that.

I've posted #533.  Would you be able to test it out to make sure it meets your needs?

**Reply by ckoehler** [reply]
Thanks so much, I'll test that out!

I guess our structs aren't secret. It looks something like this (simplified):

```
pub struct Stream<'a> {
    packets: Vec<Packet<'a>>,
    // ...
}
```

And Packet: 
```
pub struct Packet<'a> {
    backing_slice: BackingSlice<'a, u8>,
    // ...
}
```

And the backing slice holds a reference to the data. This way we don't have to copy the actual data around until we really need to. 

We have a `to_slice()` on `Stream` that returns a `Cow` with a `Borrowed` variant if there is only one `Packet`, which gives us the backing slice of that one packet; and the `Owned` variant that copies the data from all packets into one Vec if there are multiple `Packet`s and gives us that. We implemented `Iterator` etc for `Stream`, too, for ease of use. 

Anyway, let me try out your changes! Thanks again!

**Reply by ckoehler** [reply]
I replied on the PR, too: this works well! Thanks for accommodating us!
